### PR TITLE
Handle optional array inputs

### PR DESCRIPTION
**Pull request by dinithiW**
Handles optional array inputs. Added new test case.

- ~~Fix the "" defaults for non-string types~~ Moved to #80 
- [x] The arrays of files should get the .path attribute
- [x] Apply the changes from https://github.com/biowdl/tasks/pull/291/files#diff-c76c01f3ca967cdb9c157a75e7fb1a08d0037543b455c2107398601a2f526ebfR1081 to fix the file co-location error in the original WDL
- [ ] There is an extra prefix for the arrays. The separator should only be added if the current element is not the last one in the array.

### CODE REVIEW

File: wdl2cwl/main.py [hunk 1]
```diff
@@ -331,6 +331,16 @@ def get_input(
...
                     )
                 )
+            else:
+                inputs.append(
+                    cwl.CommandInputParameter(
+                        id=input_name,
+                        type=[
+                            cwl.CommandInputArraySchema(items=input_type, type="array")
+                        ],
+                        default="",
```
Comment (mr-c) [inline wdl2cwl/main.py:341]: `[]` would be a better default, as an empty string is doesn't match the type.

`defined` would also need to work differently for arrays, testing for length greater than zero instead of (in)equality with and empty string.

**Review by mr-c — CHANGES_REQUESTED** [review]

**Comment by mr-c** [comment]
Starting to get a little complicated! Can you try to find test data for this one, preferably using at least one of the optional arrays?

**Comment by dinithiW** [comment]
> Starting to get a little complicated! Can you try to find test data for this one, preferably using at least one of the optional arrays?

Any idea what kinds of files can go to intervalList?

**Comment by dinithiW** [comment]
Changed default value and JS expression for defined function

**Comment by mr-c** [comment]
> > Starting to get a little complicated! Can you try to find test data for this one, preferably using at least one of the optional arrays?
> 
> Any idea what kinds of files can go to intervalList?

https://gatk.broadinstitute.org/hc/en-us/articles/360035531852-Intervals-and-interval-lists gives some hints

**Comment by mr-c** [comment]
>     1. What should the default value be? I think we had `null` as the default value first, but changed to `""` because we didn't want null to be passed into the command.

Yeah, that made sense at the beginning of this project. Now we see that we need a new approach, so it is time to improve it.

After reading https://github.com/openwdl/wdl/blob/main/versions/1.1/SPEC.md#concatenation-of-optional-values I think that we may need to re-do how we handle optional WDL inputs. 

It seems the safer thing to do is
1. Not define a default value for optional inputs (unless the WDL document has one)

and

2. For each WDL expression placeholder `~{…}` containing references to one or more **optional** inputs
  a. replace it with CWL expression that returns an empty string if **any** of the optional inputs referenced in the WDL expression are missing (`=== null`) and otherwise returns the result of the expression itself.

Do you think this plan will work?

If you want, you can make a new issue to fix the optional WDL inputs, and then this PR can be finished by removing  temporarily removing the optional inputs from `gatk_1.wdl` (after Apply the changes from https://github.com/biowdl/tasks/pull/291/files#diff-c76c01f3ca967cdb9c157a75e7fb1a08d0037543b455c2107398601a2f526ebfR1081 to fix the file co-location error in the original WDL)

**Comment by dinithiW** [comment]
> Do you think this plan will work?
> 
> If you want, you can make a new issue to fix the optional WDL inputs, and then this PR can be finished by removing temporarily removing the optional inputs from `gatk_1.wdl` (after Apply the changes from https://github.com/biowdl/tasks/pull/291/files#diff-c76c01f3ca967cdb9c157a75e7fb1a08d0037543b455c2107398601a2f526ebfR1081 to fix the file co-location error in the original WDL)

Ok. I will create a new PR to handle the optional inputs. I have updated gatk_1.wdl after removing the optional inputs from the inputs section and the command.

File: wdl2cwl/tests/wdl_files/gatk_1.wdl [hunk 2]
```diff
@@ -34,15 +34,15 @@ task HaplotypeCaller {
         #String emitRefConfidence = if gvcf then "GVCF" else "NONE"
         Boolean dontUseSoftClippedBases = false
 
-        Array[File]+? intervalList
-        Array[File]+? excludeIntervalList
```
Comment (mr-c) [inline wdl2cwl/tests/wdl_files/gatk_1.wdl]: Let's keep these two as they are the topic of this PR (optional array inputs) and the conversion worked.

Comment (dinithiW) [reply wdl2cwl/tests/wdl_files/gatk_1.wdl]: Ok. If I apply all the changes from https://github.com/biowdl/tasks/pull/291/files#diff-c76c01f3ca967cdb9c157a75e7fb1a08d0037543b455c2107398601a2f526ebfR1081 the resultant cwl is not accurate. Shall I keep it as it is or remove the changes?

Comment (mr-c) [reply wdl2cwl/tests/wdl_files/gatk_1.wdl]: Can you show me what is not accurate about the resulting CWL?

File: wdl2cwl/tests/wdl_files/gatk_1.wdl [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        Int timeMinutes = 400 # This will likely be used with intervals, as such size based estimation can't be used.
+        String dockerImage = "quay.io/biocontainers/gatk4:4.1.8.0--py38h37ae868_0"
+    }
+
+    command {
+        set -e
+        mkdir -p "$(dirname ~{outputPath})"
+        gatk --java-options '-Xmx~{javaXmxMb}M -XX:ParallelGCThreads=1' \
+        HaplotypeCaller \
+        -R wd2/~{basename(referenceFasta)} \
```
Comment (mr-c) [inline wdl2cwl/tests/wdl_files/gatk_1.wdl:59]: Good that you copied in the `wd2/` change, but this doesn't work without the `ln -s` changes that go with it https://github.com/biowdl/tasks/pull/291/files#diff-c76c01f3ca967cdb9c157a75e7fb1a08d0037543b455c2107398601a2f526ebfR1084

File: wdl2cwl/tests/cwl_files/gatk_1.cwl [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            set -e
+            mkdir -p "\$(dirname $(inputs.outputPath))"
+            gatk --java-options '-Xmx$(inputs.javaXmxMb)M -XX:ParallelGCThreads=1' \
+            HaplotypeCaller \
+            -R wd2/$(inputs.basename(referenceFasta)) \
+            -O $(inputs.outputPath) \
+            -I ${
+            var text = "";
+            for(var i=0;i<inputs["inputBams"].length;i++) 
+              text+= inputs["inputBams"][i].path+" -I ";
```
Comment (mr-c) [inline wdl2cwl/tests/cwl_files/gatk_1.cwl:61]: This is still not working.

There is an extra prefix for the arrays. The separator should only be added if the current element is not the last one in the array.

File: wdl2cwl/tests/wdl_files/gatk_1.wdl (hunk 2 above)
Comment (dinithiW) [reply wdl2cwl/tests/wdl_files/gatk_1.wdl]: For example ```for FILE in $(inputs.sep(" ",inputBams)); do ln -s $FILE wd/\$(inputBams $FILE) ; done```

File: wdl2cwl/tests/wdl_files/gatk_1.wdl (hunk 3 above)
Comment (dinithiW) [reply wdl2cwl/tests/wdl_files/gatk_1.wdl:59]: Yes. I only saw those changes later :/

File: wdl2cwl/tests/cwl_files/gatk_1.cwl (hunk 4 above)
Comment (dinithiW) [reply wdl2cwl/tests/cwl_files/gatk_1.cwl]: I modified the code to address this
